### ISSUE TITLE
html5 client - High CPU load on client side while receiving incoming video from other call participants

### ISSUE THREAD

**Issue by Azbests** [bigbluebutton/bigbluebutton#8897]
What is expected CPU load on client when using BBB?

I have encountered problem that people get high cpu loads on their side when even a
single incoming video/audio stream is active. With session where nine people tried to
use BBB for video conference poor laptops were pretty much running 100% cpu load.

At the moment I haven't found anything on web that would point to explanation for this
high cpu load on client side or solution for that issue.

Any comments? 

Clients: Various Chrome and Firefox newer versions on mostly Win10 hosts, some macs.
On server side I see minimal load, as expected.

Interestingly when testing with my workstation (to powerful and to high core count cpu, probably) or my older 13* macbook 2015, I do not see that big of a load. Mac gets cpu load in ranges of 30-65% w/ Chrome as browser

BBB ver:  2.2.2 (1781) w/ HTML5 client only config - Installed the manual way as was instructed in docs install article.
Host OS ver: Ubuntu 16.04.6 LTS

**Comment by GhaziTriki** [comment]
@Azbests Could please update to 2.2.3, do `sudo bbb-conf --restart` and tell us if you still see have this issue?

**Comment by usaafko** [comment]
Have same problem on 2.2.3 version. Same install as @Azbests 
My laptop has i3-2410M processor and 4Gb RAM, when I join meeting with ~ 9 people on video (Chrome 80, Windows 10): 100%CPU load and high fan speed

What can it be? Zoom client don't have such load on my system

**Comment by Azbests** [comment]
Hmm, interesting. I do not see v2.2.3 provided by apt repo

```
bbb-apps/unknown 1:2.2.0-42 amd64 [upgradable from: 1:2.2.0-39]
bbb-apps-akka/unknown,unknown 2.2.0-77 all [upgradable from: 2.2.0-73]
bbb-apps-screenshare/unknown 1:2.2.0-42 amd64 [upgradable from: 1:2.2.0-39]
bbb-apps-video/unknown 1:2.2.0-39 amd64 [upgradable from: 1:2.2.0-36]
bbb-config/unknown 1:2.2.0-170 amd64 [upgradable from: 1:2.2.0-160]
bbb-etherpad/unknown 1:2.2.0-50 amd64 [upgradable from: 1:2.2.0-48]
bbb-freeswitch-core/unknown 2:2.2.0-97 amd64 [upgradable from: 2:2.2.0-90]
bbb-fsesl-akka/unknown,unknown 2.2.0-61 all [upgradable from: 2.2.0-58]
bbb-html5/unknown 1:2.2.0-870 amd64 [upgradable from: 1:2.2.0-862]
bbb-record-core/unknown 1:2.2.0-62 amd64 [upgradable from: 1:2.2.0-60]
bbb-web/unknown 1:2.2.0-117 amd64 [upgradable from: 1:2.2.0-114]
bbb-webrtc-sfu/unknown 1:2.2.0-93 amd64 [upgradable from: 1:2.2.0-89]
```
```bash
root@bbb:~# cat /etc/apt/sources.list.d/bigbluebutton.list 
deb https://ubuntu.bigbluebutton.org/xenial-220/ bigbluebutton-xenial main
root@bbb:~# cat /etc/apt/sources.list.d/bigbluebutton-ubuntu-support-xenial.list 
deb http://ppa.launchpad.net/bigbluebutton/support/ubuntu xenial main
# deb-src http://ppa.launchpad.net/bigbluebutton/support/ubuntu xenial main
```

**Comment by prlanzarin** [comment]
@Azbests Please look at the docs on how to reduce the bitrate settings for webcams https://docs.bigbluebutton.org/2.2/customize.html#reduce-bandwidth-for-webcams.
You can go as low as 30kbps. The lower, the less CPU and bandwidth usage you'll get.
Also look at the "See other viewers webcams" lock setting, which reduces the number of subscribing streams viewers get.

We also have some optimization problems on the camera UI in 2.2, but I don't think they'd the cause of a CPU hog and they'll probably only be addressed in 2.3.

**Comment by Azbests** [comment]
I did take a look at those settings before, but I thought that it was more for server side of things in case if it cannot handle load or network bandwidth is an issue?

Either way, will take a look what changes if I reduce the quality via these settings.

**Comment by mreichh** [comment]
Hi all,

we are experiencing high/maxing out client-side CPU load as well. I think this is because of the SFU architecture of BBB's video transmission.

Made a couple of tests during the last days to find out how BBB scales when it comes to bigger meetings where all participants need to see each other all the time. Tested on a fresh BBB 2.2.3 instance, created by bbb-install.sh and hosted by a machine with 36 cores, 48GB of RAM, 2G ethernet and SSD storage, it came out pretty clearly that meetings with more than 20 audio-webcam participants start getting shaky for clients with middle-low network or hardware resources. Meetings with more than 25 audio-webcam connections are only feasible if the hardware and network situation on the client-side is very good. Starting from 30 participants, meetings including audio and webcam for everyone are not realistic anymore - no way for tablets oder below i7 notebooks withouth dedicated graphics to attend such meetings.

In addition reducing the video bitrate from 200 to 100kbit/sec did not lead and meaningful increase of max. user count for an audio-video meeting. In fact this bitrate reduction only provided a small user count increase of around 10%. So obviously there are other limitting factors for big conferences than just the video bitrate. BTW: The server-side resource consumption, network- and CPU-wise, was trivial resp. at 15% max. during our tests.

We are pretty happy with the features and performance of BBB for groups below 20 participants. But, there's a huge demand as well for bigger audio-video conferences of like 30-50 people. Though, such meetings are practically impossible within BBB, even if the video bitrate is reduced to 100kbit/sec.

Unfortunately big audio-video meetings are a killer feature which products like Zoom or MS Teams provide as they use a MCU architecture for video transmission (afaik).

So, may I ask if there are other BBB parameters than the video bitrate which could be tweaked to handle the CPU load problem on the clients and to support more than 30 audio-video participants per meeting? Like an option for a dynamic bitrate adjustment on the client-side or so?

Or is there a plan to implement an MCU-like video transmission which can be optionally activated if audio-video meetings get bigger than 25 people? I mean hardware for big MCU servers wouldn't be the biggest cost driver anymore, these days.

It'd be really great if BBB could also ne used to host big conferences as this is crucial for many scenarios like interactive seminars at universities, big committee meetings or big distributed work groups. 

Thanks in advance for your feedback!

**Comment by Azbests** [comment]
With 
* Lowered quality settings for webcams
```
    cameraProfiles:
    - id: low
      name: Low quality
      default: true
      bitrate: 150
    - id: medium
      name: Medium quality
      default: false
      bitrate: 200
    - id: high
      name: High quality
      default: false
      bitrate: 250
    - id: hd
      name: High definition
      default: false
      bitrate: 400
```
* Running v2.2.3
* 10 Participants with webcams
* Task-man screencap from one of participants laptops
<img width="510" alt="Screenshot_2" src="https://user-images.githubusercontent.com/13135279/79962397-a95fb400-8490-11ea-86a1-4d30cd2a7ebf.png">

Anything new in v2.2.4/v2.2.5 that could change things or as noted before some changes are to be expected at v2.3 release?

Will update anyhow, but first need to write script to return settings and modified html files to our preferences after update.

**Comment by mreichh** [comment]
already updated to 2.2.5 but is does not seem to bring any improvements in this regard. Actually the issue gets even bigger in a common scenario where one participant - normally the speaker, which is the most important person to be seen and unterstood - might additionally use screensharing.

**Comment by iSamof** [comment]
@Azbests - As a side note, I think I read somewhere that in the later versions of BBB the settings are kept in a custom file and are not touched by an update. If you are aware of this, please can you share  your reasons for writing a script to keep settings?

**Comment by beatwiz** [comment]
Im forcing Low Profiles with 80 bitrate and the usage is still high. In some older laptops everything goes slowmotion and eventually timeouts occur and people start to drop from the conference. I totally understand that we cant support all the ancient hardware but this is noticeable even on modern hardware.

**Comment by Azbests** [comment]
@iSamof I wasn't aware of that, no. 

But I think that when upgrading to v2.2.3 there were some changes in `/usr/share/meteor/bundle/programs/server/assets/app/config/settings.yml` after upgrade, not done by me. 

As far as I can remember at least `external video support` feature was re-enabled, which I had disabled previously.

I will check again when upgrading to latest version from repo.

I anyhow need the script as I had modified default html files. But those I guess were expected to be rewritten by update. I just hadn't saved my modified files to re-apply them after the update.
